aip/client-libraries/4290.md

```diff
@@ -0,0 +1,202 @@
+arguments, and this should be a last resort if they are unavoidable.
+
+**Note:** Even though each component of this is standard in the Docker
+ecosystem (other than the destination path issue, which is a result of how
+protoc handles imports), this is still a rather long command. We can provide a
+[shortcut script][] to further simplify this, but such a script would be for
+convenience and not a replacement for this interface.
+
+### Container composition
+
+Containers **must** include:
+
+- A current version of protoc, the protocol buffer compiler.
+- Common protos permitted to be used by all APIs ([api-common-protos][]).
+- The applicable code generator plugin, as well as any dependencies it
+  requires.
+  - The code generator plugin itself **should** be added using an `ADD` or
+    `COPY` statement from the host machine at build time and installed locally;
+    it **should not** pull from a package manager. (This leads to catch-22
+    situations when cutting releases.)
+  - Images **may** include either a pre-compiled binary of the plugin, or the
+    installed source code, depending on the needs of the applicable ecosystem.
+  - Installation of dependencies **should** use appropriate package managers.
+
+The common protos and the protoc compiler are supplied by an independent image
+(`gcr.io/gapic-images/api-common-protos`). Both `protoc` and the common protos
+can be retrieved from this image into a generator's image using the
+`COPY --from` syntax (see [multi-stage builds][]). This is the preferred
+approach as it follows Docker conventions, and allows the protos to be
+versioned independently.
+
+### Base images
+
+**TL;DR:** ach language probably wants `language:x.y-alpine` or
+`language:x.y-slim`. For example, `ruby:2.5-alpine` or `python:3.7-slim`.
+(Alpine images are smaller but idiosyncratic.)
+
+The following guidelines apply to selecting base images (sorted roughly from
+most important to least important):
+
+- Images **should** generally be based off an official image for the latest
+  stable version of the language in which the generator is implemented.
+- Images **should** be able to install required system dependencies from a
+  well-understood package manager.
+- Images **should** be ultimately based off of Alpine, Debian, or Ubuntu. This
+  is to ensure we benefit from GCR's [vulnerability scanning][].
+- Images should endeavor to be as small as possible, in line with the general
+  expectations of the Docker community:
+  - Use the smallest base image you can. Alpine-based images are great if
+    possible, but may not always be reasonable. "Slim" Debian images are
+    usually the next best (and probably significantly more feasible in many
+    situations).
+
+### Mount points
+
+`protoc` must read protos (representing the API to be generated) from disk, and
+must write the final output (the client library) to disk. Because the user has
+the API protos on his or her host machine, and will ultimately need the output
```

Comment:
`his or her` -> `their`, gender isn't exactly necessary here.

Reply:
Not a fan of "singular their", but "the" works here and has the same effect.